// @rhinestone/module-sdk/_cjs/module/hook-multi-plexer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HookType = exports.getHooks = exports.getRemoveHookAction = exports.getAddHookAction = exports.getHookMultiPlexer = exports.HOOK_MULTI_PLEXER_ADDRESS = void 0;
var constants_1 = require("./constants");
Object.defineProperty(exports, "HOOK_MULTI_PLEXER_ADDRESS", { enumerable: true, get: function () { return constants_1.HOOK_MULTI_PLEXER_ADDRESS; } });
var installation_1 = require("./installation");
Object.defineProperty(exports, "getHookMultiPlexer", { enumerable: true, get: function () { return installation_1.getHookMultiPlexer; } });
var usage_1 = require("./usage");
Object.defineProperty(exports, "getAddHookAction", { enumerable: true, get: function () { return usage_1.getAddHookAction; } });
Object.defineProperty(exports, "getRemoveHookAction", { enumerable: true, get: function () { return usage_1.getRemoveHookAction; } });
Object.defineProperty(exports, "getHooks", { enumerable: true, get: function () { return usage_1.getHooks; } });
var types_1 = require("./types");
Object.defineProperty(exports, "HookType", { enumerable: true, get: function () { return types_1.HookType; } });
