// @rhinestone/module-sdk/_cjs/module/hook-multi-plexer/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHookMultiPlexer = void 0;
const viem_1 = require("viem");
const constants_1 = require("../../constants");
const getHookMultiPlexer = ({ globalHooks, valueHooks, delegatecallHooks, sigHooks, targetHooks, selector, hookType, }) => {
    return {
        address: constants_1.GLOBAL_CONSTANTS.HOOK_MULTI_PLEXER_ADDRESS,
        module: constants_1.GLOBAL_CONSTANTS.HOOK_MULTI_PLEXER_ADDRESS,
        initData: (0, viem_1.encodeAbiParameters)([
            { internalType: 'address[]', name: 'globalHooks', type: 'address[]' },
            { internalType: 'address[]', name: 'valueHooks', type: 'address[]' },
            {
                internalType: 'address[]',
                name: 'delegatecallHooks',
                type: 'address[]',
            },
            {
                components: [
                    { internalType: 'address[]', name: 'subHooks', type: 'address[]' },
                    { internalType: 'bytes4', name: 'sig', type: 'bytes4' },
                ],
                name: 'sigHooks',
                type: 'tuple[]',
            },
            {
                components: [
                    { internalType: 'address[]', name: 'subHooks', type: 'address[]' },
                    { internalType: 'bytes4', name: 'sig', type: 'bytes4' },
                ],
                name: 'targetHooks',
                type: 'tuple[]',
            },
        ], [globalHooks, valueHooks, delegatecallHooks, sigHooks, targetHooks]),
        deInitData: '0x',
        additionalContext: '0x',
        type: 'hook',
        selector,
        hookType,
    };
};
exports.getHookMultiPlexer = getHookMultiPlexer;
