// @rhinestone/module-sdk/_esm/common/queries/account.js
import { GLOBAL_CONSTANTS } from '../../constants.js';
const query = `
    query ($smartAccount: String, $chainId: Int) {
       SmartAccount_ModuleQuery (where: { smartAccount: { _eq: $smartAccount }, isInstalled: { _eq: true }, chainId: { _eq: $chainId }}) {
        moduleAddress,
        moduleTypeId
      }
    }
  `;
export const getInstalledModules = async ({ account, client, }) => {
    const variables = {
        smartAccount: account.address,
        chainId: await client.getChainId(),
    };
    const response = await fetch(GLOBAL_CONSTANTS.INDEXER_URL, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            query,
            variables,
        }),
    });
    const responseBody = await response.json();
    if (response.ok) {
        return responseBody.data.SmartAccount_ModuleQuery.map((module) => module.module);
    }
    else {
        throw new Error(`Error: ${responseBody.errors
            .map((error) => error.message)
            .join(', ')}`);
    }
};
