// @rhinestone/module-sdk/_cjs/module/zk-email-recovery/universal-email-recovery/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUniversalEmailRecoveryExecutor = void 0;
const viem_1 = require("viem");
const constants_1 = require("../../../constants");
const getUniversalEmailRecoveryExecutor = ({ validator, isInstalledContext, initialSelector, guardians, weights, threshold, delay, expiry, chainId, hook, }) => {
    return {
        address: constants_1.GLOBAL_CONSTANTS.UNIVERSAL_EMAIL_RECOVERY_ADDRESS,
        module: constants_1.GLOBAL_CONSTANTS.UNIVERSAL_EMAIL_RECOVERY_ADDRESS,
        initData: (0, viem_1.encodeAbiParameters)([
            { name: 'validator', type: 'address' },
            { name: 'isInstalledContext', type: 'bytes' },
            { name: 'initialSelector', type: 'bytes4' },
            { name: 'guardians', type: 'address[]' },
            { name: 'weights', type: 'uint256[]' },
            { name: 'delay', type: 'uint256' },
            { name: 'expiry', type: 'uint256' },
            { name: 'threshold', type: 'uint256' },
        ], [
            validator,
            isInstalledContext,
            initialSelector,
            guardians,
            weights.map((weight) => BigInt(weight)),
            BigInt(threshold),
            BigInt(delay),
            BigInt(expiry),
        ]),
        deInitData: '0x',
        additionalContext: '0x',
        type: 'executor',
        hook,
    };
};
exports.getUniversalEmailRecoveryExecutor = getUniversalEmailRecoveryExecutor;
