// @rhinestone/module-sdk/_esm/account/safe/api/uninstallModule.js
import { encodeAbiParameters, encodeFunctionData, parseAbi, parseAbiParameters, } from 'viem';
import { isModuleInstalled } from './isModuleInstalled.js';
import { accountAbi } from '../constants/abis.js';
import { moduleTypeIds } from '../../../module/types.js';
import { getPreviousModule } from '../../../common/getPrevModule.js';
export const uninstallModule = ({ client, account, module, }) => {
    switch (module.type) {
        case 'validator':
        case 'executor':
            return _uninstallModule({ client, account, module });
        case 'hook':
            if (!module.selector || module.hookType === undefined) {
                throw new Error(`hookType and selector params are required for module type ${module.type}`);
            }
            return _uninstallModule({ client, account, module });
        case 'fallback':
            if (!module.functionSig) {
                throw new Error(`functionSig param is required for module type ${module.type}`);
            }
            return _uninstallModule({ client, account, module });
        default:
            throw new Error(`Unknown module type ${module.type}`);
    }
};
const _uninstallModule = async ({ client, account, module, }) => {
    const executions = [];
    const isInstalled = await isModuleInstalled({ client, account, module });
    if (isInstalled) {
        let moduleData = getModuleCalldata(module);
        if (module.type === 'validator' || module.type === 'executor') {
            const prev = await getPreviousModule({ client, account, module });
            moduleData = encodeAbiParameters([
                { name: 'prev', type: 'address' },
                { name: 'moduleInitData', type: 'bytes' },
            ], [prev, moduleData]);
        }
        const data = encodeFunctionData({
            functionName: 'uninstallModule',
            abi: parseAbi(accountAbi),
            args: [BigInt(moduleTypeIds[module.type]), module.module, moduleData],
        });
        executions.push({
            to: account.address,
            target: account.address,
            value: BigInt(0),
            callData: data,
            data,
        });
    }
    return executions;
};
const getModuleCalldata = (module) => {
    switch (module.type) {
        case 'validator':
        case 'executor':
            return module.deInitData;
        case 'hook':
            return encodeAbiParameters(parseAbiParameters('uint8 hookType, bytes4 selector, bytes memory deInitData'), [module.hookType, module.selector, module.deInitData]);
        case 'fallback':
            return encodeAbiParameters(parseAbiParameters('bytes4 functionSig, bytes memory moduleDeInitData'), [module.functionSig, module.deInitData]);
        default:
            throw new Error(`Unknown module type ${module.type}`);
    }
};
