// @rhinestone/module-sdk/_cjs/module/smart-sessions/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSmartSessionsCompatibilityFallback = exports.getSmartSessionsValidator = void 0;
const viem_1 = require("viem");
const abi_1 = require("./abi");
const types_1 = require("./types");
const types_2 = require("../../module/types");
const constants_1 = require("../../constants");
const getSmartSessionsValidator = ({ sessions, useRegistry = true, hook, }) => {
    return {
        address: constants_1.GLOBAL_CONSTANTS.SMART_SESSIONS_ADDRESS,
        module: constants_1.GLOBAL_CONSTANTS.SMART_SESSIONS_ADDRESS,
        initData: sessions?.length
            ? (0, viem_1.encodePacked)(['bytes1', 'bytes'], [
                useRegistry
                    ? types_1.SmartSessionMode.ENABLE
                    : types_1.SmartSessionMode.UNSAFE_ENABLE,
                (0, viem_1.encodeAbiParameters)(abi_1.installSmartSessionsAbi, [sessions]),
            ])
            : '0x',
        deInitData: '0x',
        additionalContext: '0x',
        type: 'validator',
        hook,
    };
};
exports.getSmartSessionsValidator = getSmartSessionsValidator;
const getSmartSessionsCompatibilityFallback = () => {
    return {
        address: constants_1.GLOBAL_CONSTANTS.SMART_SESSIONS_COMPATIBILITY_FALLBACK_ADDRESS,
        module: constants_1.GLOBAL_CONSTANTS.SMART_SESSIONS_COMPATIBILITY_FALLBACK_ADDRESS,
        initData: '0x',
        deInitData: '0x',
        selector: '0x84b0196e',
        callType: types_2.CallType.CALLTYPE_SINGLE,
        additionalContext: '0x',
        type: 'fallback',
    };
};
exports.getSmartSessionsCompatibilityFallback = getSmartSessionsCompatibilityFallback;
