// @rhinestone/module-sdk/_esm/module/scheduled-transfers/usage.js
import { encodeAbiParameters, encodeFunctionData, encodePacked, zeroAddress, } from 'viem';
import { abi } from './abi.js';
import { GLOBAL_CONSTANTS } from '../../constants.js';
export const getScheduledTransferData = ({ scheduledTransfer, }) => {
    const amount = BigInt(Number(scheduledTransfer.amount) *
        10 ** (scheduledTransfer.token?.decimals || 18));
    return encodeAbiParameters([
        { name: 'recipient', type: 'address' },
        { name: 'token', type: 'address' },
        { name: 'amount', type: 'uint256' },
    ], [
        scheduledTransfer.recipient,
        scheduledTransfer.token
            ? scheduledTransfer.token.token_address
            : zeroAddress,
        amount,
    ]);
};
export const getCreateScheduledTransferAction = ({ scheduledTransfer, }) => {
    const data = encodeFunctionData({
        functionName: 'addOrder',
        abi,
        args: [
            encodePacked(['uint48', 'uint16', 'uint48', 'bytes'], [
                scheduledTransfer.repeatEvery,
                scheduledTransfer.numberOfRepeats,
                scheduledTransfer.startDate,
                getScheduledTransferData({ scheduledTransfer }),
            ]),
        ],
    });
    return {
        to: GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        target: GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
export const getExecuteScheduledTransferAction = ({ jobId, }) => {
    const data = encodeFunctionData({
        functionName: 'executeOrder',
        abi,
        args: [BigInt(jobId)],
    });
    return {
        to: GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        target: GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
export const getToggleScheduledTransferAction = ({ jobId, }) => {
    const data = encodeFunctionData({
        functionName: 'toggleOrder',
        abi,
        args: [BigInt(jobId)],
    });
    return {
        to: GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        target: GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
