// @rhinestone/module-sdk/_cjs/module/omni-account/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccountLockerTargetExecutor = exports.getAccountLockerSourceExecutor = exports.getAccountLockerHook = void 0;
const viem_1 = require("viem");
const constants_1 = require("../../constants");
const getAccountLockerHook = ({ isOmniMode, hook } = {
    isOmniMode: false,
}) => {
    return {
        module: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_HOOK,
        address: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_HOOK,
        initData: (0, viem_1.encodeAbiParameters)([{ name: 'enableOmniLock', type: 'bool' }], [isOmniMode]),
        deInitData: '0x',
        additionalContext: '0x',
        hook,
        type: 'hook',
    };
};
exports.getAccountLockerHook = getAccountLockerHook;
const getAccountLockerSourceExecutor = ({ hook, } = {}) => {
    return {
        module: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        address: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        initData: '0x',
        deInitData: '0x',
        additionalContext: '0x',
        hook,
        type: 'executor',
    };
};
exports.getAccountLockerSourceExecutor = getAccountLockerSourceExecutor;
const getAccountLockerTargetExecutor = ({ hook, moduleType, } = { moduleType: 'executor' }) => {
    return {
        module: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_TARGET_EXECUTOR,
        address: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_TARGET_EXECUTOR,
        initData: '0x',
        deInitData: '0x',
        additionalContext: '0x',
        hook,
        type: moduleType,
    };
};
exports.getAccountLockerTargetExecutor = getAccountLockerTargetExecutor;
