// @rhinestone/module-sdk/_cjs/module/omni-account/usage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRegisterApprovalSpendAction = exports.getDepositToAcrossAction = exports.getUnlockFundsAction = void 0;
const viem_1 = require("viem");
const abis_1 = require("./abis");
const constants_1 = require("../../constants");
const getUnlockFundsAction = ({ orchestratorSignature, request, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'unlockFundsForAccount',
        abi: abis_1.accountLockerSourceExecutorAbi,
        args: [orchestratorSignature, request],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        target: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        value: BigInt(0),
        data,
        callData: data,
    };
};
exports.getUnlockFundsAction = getUnlockFundsAction;
const getDepositToAcrossAction = ({ originModulePayload, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'handleAcross',
        abi: abis_1.accountLockerSourceExecutorAbi,
        args: [originModulePayload],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        target: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        value: BigInt(0),
        data,
        callData: data,
    };
};
exports.getDepositToAcrossAction = getDepositToAcrossAction;
const getRegisterApprovalSpendAction = ({ approvalSpend, orchestratorSignature, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'registerApprovalSpend',
        abi: abis_1.accountLockerSourceExecutorAbi,
        args: [approvalSpend, orchestratorSignature],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        target: constants_1.GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        value: BigInt(0),
        data,
        callData: data,
    };
};
exports.getRegisterApprovalSpendAction = getRegisterApprovalSpendAction;
