// @rhinestone/module-sdk/_cjs/module/auto-savings/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAutoSavingsExecutor = void 0;
const viem_1 = require("viem");
const uniswap_1 = require("../utils/uniswap");
const constants_1 = require("../../constants");
const getAutoSavingsExecutor = ({ chainId, configs, hook, }) => {
    const swapRouter = uniswap_1.swapRouterAddresses[chainId];
    if (!swapRouter) {
        throw new Error(`Swap router address not found for chainId ${chainId}`);
    }
    return {
        address: constants_1.GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
        module: constants_1.GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
        initData: (0, viem_1.encodeAbiParameters)([
            { internalType: 'address', name: 'swapRouter', type: 'address' },
            {
                components: [
                    { internalType: 'address', name: 'token', type: 'address' },
                    { internalType: 'uint64', name: 'percentage', type: 'uint64' },
                    { internalType: 'address', name: 'vault', type: 'address' },
                ],
                internalType: 'struct AutoSavings.ConfigWithToken[]',
                name: '_configs',
                type: 'tuple[]',
            },
        ], [swapRouter, configs]),
        deInitData: '0x',
        additionalContext: '0x',
        type: 'executor',
        hook,
    };
};
exports.getAutoSavingsExecutor = getAutoSavingsExecutor;
