// @rhinestone/module-sdk/_cjs/module/deadman-switch/usage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDeadmanSwitchValidatorMockSignature = exports.getDeadmanSwitchConfig = void 0;
const abi_1 = require("./abi");
const constants_1 = require("../../constants");
const getDeadmanSwitchConfig = async ({ account, client, }) => {
    try {
        const config = (await client.readContract({
            address: constants_1.GLOBAL_CONSTANTS.DEADMAN_SWITCH_ADDRESS,
            abi: abi_1.abi,
            functionName: 'config',
            args: [account.address],
        }));
        return config;
    }
    catch (err) {
        throw new Error('Error getting config');
    }
};
exports.getDeadmanSwitchConfig = getDeadmanSwitchConfig;
const getDeadmanSwitchValidatorMockSignature = () => {
    return '0xe8b94748580ca0b4993c9a1b86b5be851bfc076ff5ce3a1ff65bf16392acfcb800f9b4f1aef1555c7fce5599fffb17e7c635502154a0333ba21f3ae491839af51c';
};
exports.getDeadmanSwitchValidatorMockSignature = getDeadmanSwitchValidatorMockSignature;
