// @rhinestone/module-sdk/_esm/module/smart-sessions/policies/universal-action-policy/installation.js
import { MAX_RULES } from './constants.js';
import { encodeAbiParameters } from 'viem';
import { GLOBAL_CONSTANTS } from '../../../../constants.js';
export const getUniversalActionPolicy = (actionConfig) => {
    if (actionConfig.paramRules.rules.length > MAX_RULES) {
        throw new Error(`Max number of rules is ${MAX_RULES}`);
    }
    return {
        policy: GLOBAL_CONSTANTS.UNIVERSAL_ACTION_POLICY_ADDRESS,
        address: GLOBAL_CONSTANTS.UNIVERSAL_ACTION_POLICY_ADDRESS,
        initData: encodeAbiParameters([
            {
                components: [
                    {
                        name: 'valueLimitPerUse',
                        type: 'uint256',
                    },
                    {
                        components: [
                            {
                                name: 'length',
                                type: 'uint256',
                            },
                            {
                                components: [
                                    {
                                        name: 'condition',
                                        type: 'uint8',
                                    },
                                    {
                                        name: 'offset',
                                        type: 'uint64',
                                    },
                                    {
                                        name: 'isLimited',
                                        type: 'bool',
                                    },
                                    {
                                        name: 'ref',
                                        type: 'bytes32',
                                    },
                                    {
                                        components: [
                                            {
                                                name: 'limit',
                                                type: 'uint256',
                                            },
                                            {
                                                name: 'used',
                                                type: 'uint256',
                                            },
                                        ],
                                        name: 'usage',
                                        type: 'tuple',
                                    },
                                ],
                                name: 'rules',
                                type: 'tuple[16]',
                            },
                        ],
                        name: 'paramRules',
                        type: 'tuple',
                    },
                ],
                name: 'ActionConfig',
                type: 'tuple',
            },
        ], [actionConfig]),
    };
};
