// @rhinestone/module-sdk/_esm/module/omni-account/installation.js
import { encodeAbiParameters } from 'viem';
import { GLOBAL_CONSTANTS } from '../../constants.js';
export const getAccountLockerHook = ({ isOmniMode, hook } = {
    isOmniMode: false,
}) => {
    return {
        module: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_HOOK,
        address: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_HOOK,
        initData: encodeAbiParameters([{ name: 'enableOmniLock', type: 'bool' }], [isOmniMode]),
        deInitData: '0x',
        additionalContext: '0x',
        hook,
        type: 'hook',
    };
};
export const getAccountLockerSourceExecutor = ({ hook, } = {}) => {
    return {
        module: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        address: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        initData: '0x',
        deInitData: '0x',
        additionalContext: '0x',
        hook,
        type: 'executor',
    };
};
export const getAccountLockerTargetExecutor = ({ hook, moduleType, } = { moduleType: 'executor' }) => {
    return {
        module: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_TARGET_EXECUTOR,
        address: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_TARGET_EXECUTOR,
        initData: '0x',
        deInitData: '0x',
        additionalContext: '0x',
        hook,
        type: moduleType,
    };
};
