// @rhinestone/module-sdk/_esm/module/omni-account/usage.js
import { encodeFunctionData } from 'viem';
import { accountLockerSourceExecutorAbi } from './abis/index.js';
import { GLOBAL_CONSTANTS } from '../../constants.js';
export const getUnlockFundsAction = ({ orchestratorSignature, request, }) => {
    const data = encodeFunctionData({
        functionName: 'unlockFundsForAccount',
        abi: accountLockerSourceExecutorAbi,
        args: [orchestratorSignature, request],
    });
    return {
        to: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        target: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        value: BigInt(0),
        data,
        callData: data,
    };
};
export const getDepositToAcrossAction = ({ originModulePayload, }) => {
    const data = encodeFunctionData({
        functionName: 'handleAcross',
        abi: accountLockerSourceExecutorAbi,
        args: [originModulePayload],
    });
    return {
        to: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        target: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        value: BigInt(0),
        data,
        callData: data,
    };
};
export const getRegisterApprovalSpendAction = ({ approvalSpend, orchestratorSignature, }) => {
    const data = encodeFunctionData({
        functionName: 'registerApprovalSpend',
        abi: accountLockerSourceExecutorAbi,
        args: [approvalSpend, orchestratorSignature],
    });
    return {
        to: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        target: GLOBAL_CONSTANTS.ACCOUNT_LOCKER_SOURCE_EXECUTOR,
        value: BigInt(0),
        data,
        callData: data,
    };
};
