// @rhinestone/module-sdk/_cjs/account/erc7579-implementation/api/installModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.installModule = void 0;
const viem_1 = require("viem");
const isModuleInstalled_1 = require("./isModuleInstalled");
const types_1 = require("../../../module/types");
const abis_1 = require("../constants/abis");
const installModule = ({ client, account, module, }) => {
    switch (module.type) {
        case 'validator':
        case 'executor':
        case 'hook':
            return _installModule({ client, account, module });
        case 'fallback':
            if (!module.selector || !module.callType) {
                throw new Error(`Selector and callType params are required for module type ${module.type}`);
            }
            return installFallback({ client, account, module });
        default:
            throw new Error(`Unknown module type ${module.type}`);
    }
};
exports.installModule = installModule;
const _installModule = async ({ client, account, module, }) => {
    const executions = [];
    const isInstalled = await (0, isModuleInstalled_1.isModuleInstalled)({ client, account, module });
    if (!isInstalled) {
        const data = (0, viem_1.encodeFunctionData)({
            functionName: 'installModule',
            abi: (0, viem_1.parseAbi)(abis_1.accountAbi),
            args: [
                BigInt(types_1.moduleTypeIds[module.type]),
                module.module,
                module.initData || '0x',
            ],
        });
        executions.push({
            to: account.address,
            target: account.address,
            value: BigInt(0),
            callData: data,
            data,
        });
    }
    return executions;
};
async function installFallback({ client, account, module, }) {
    const executions = [];
    const isInstalled = await (0, isModuleInstalled_1.isModuleInstalled)({
        client,
        account,
        module,
    });
    if (!isInstalled) {
        const data = (0, viem_1.encodeFunctionData)({
            functionName: 'installModule',
            abi: (0, viem_1.parseAbi)(abis_1.accountAbi),
            args: [
                BigInt(types_1.moduleTypeIds[module.type]),
                module.module,
                (0, viem_1.encodePacked)(['bytes4', 'bytes1', 'bytes'], [module.selector, module.callType, module.initData ?? '0x']),
            ],
        });
        executions.push({
            to: account.address,
            target: account.address,
            value: BigInt(0),
            callData: data,
            data,
        });
    }
    return executions;
}
