// @rhinestone/module-sdk/_cjs/module/registry/usage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findTrustedAttesters = exports.getTrustAttestersAction = exports.fetchRegistryModules = void 0;
const viem_1 = require("viem");
const queries_1 = require("../../common/queries");
const abi_1 = require("./abi");
const constants_1 = require("../../constants");
const fetchRegistryModules = ({ client }) => (0, queries_1.getRegistryModules)({ client });
exports.fetchRegistryModules = fetchRegistryModules;
const getTrustAttestersAction = ({ attesters, threshold, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'trustAttesters',
        abi: abi_1.abi,
        args: [threshold, attesters],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.REGISTRY_ADDRESS,
        target: constants_1.GLOBAL_CONSTANTS.REGISTRY_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
exports.getTrustAttestersAction = getTrustAttestersAction;
const findTrustedAttesters = async ({ client, accountAddress, }) => {
    return (await client.readContract({
        address: constants_1.GLOBAL_CONSTANTS.REGISTRY_ADDRESS,
        abi: abi_1.abi,
        functionName: 'findTrustedAttesters',
        args: [accountAddress],
    }));
};
exports.findTrustedAttesters = findTrustedAttesters;
