// @rhinestone/module-sdk/_cjs/common/queries/registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRegistryModules = void 0;
const constants_1 = require("../../constants");
const query = `
  query ($chainId: Int) {
     Registry_ModuleRegistration (where: { chainId: { _eq: $chainId } }) {
        id
        implementation
      }
  }
`;
const getRegistryModules = async ({ client, }) => {
    const variables = {
        chainId: await client.getChainId(),
    };
    const response = await fetch(constants_1.GLOBAL_CONSTANTS.INDEXER_URL, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            query,
            variables,
        }),
    });
    const responseBody = await response.json();
    if (response.ok) {
        return responseBody.data.Registry_ModuleRegistration;
    }
    else {
        throw new Error(`Error: ${responseBody.errors
            .map((error) => error.message)
            .join(', ')}`);
    }
};
exports.getRegistryModules = getRegistryModules;
