// @rhinestone/module-sdk/_esm/module/auto-savings/usage.js
import { encodeFunctionData, getAddress } from 'viem';
import { GLOBAL_CONSTANTS } from '../../constants.js';
import { abi } from './abi.js';
import { SENTINEL_ADDRESS } from '../../common/constants.js';
import { getSwapDetails } from '../utils/uniswap.js';
export const getSetAutoSavingConfigAction = ({ token, config, }) => {
    const data = encodeFunctionData({
        functionName: 'setConfig',
        abi,
        args: [token, config],
    });
    return {
        to: GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
        target: GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
export const getAutoSavingTokens = async ({ account, client, }) => {
    try {
        const tokens = (await client.readContract({
            address: GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
            abi,
            functionName: 'getTokens',
            args: [account.address],
        }));
        return tokens;
    }
    catch (err) {
        console.error(err);
        return [];
    }
};
export const getDeleteAutoSavingConfigAction = async ({ account, client, token, }) => {
    try {
        const allTokens = await getAutoSavingTokens({ account, client });
        let prevToken;
        const currentTokenIndex = allTokens.findIndex((t) => t === token);
        if (currentTokenIndex === -1) {
            throw new Error('Token not found');
        }
        else if (currentTokenIndex === 0) {
            prevToken = SENTINEL_ADDRESS;
        }
        else {
            prevToken = getAddress(allTokens[currentTokenIndex - 1]);
        }
        const data = encodeFunctionData({
            functionName: 'deleteConfig',
            abi,
            args: [prevToken, token],
        });
        return {
            to: GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
            target: GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
            value: BigInt(0),
            callData: data,
            data,
        };
    }
    catch {
        throw new Error(`Failed to delete config for token ${token}`);
    }
};
export const getAutoSaveAction = ({ token, amountReceived, }) => {
    const swapDetails = getSwapDetails();
    const data = encodeFunctionData({
        functionName: 'autoSave',
        abi,
        args: [
            token,
            BigInt(amountReceived),
            swapDetails.sqrtPriceLimitX96,
            swapDetails.amountOutMin,
            swapDetails.fee,
        ],
    });
    return {
        to: GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
        target: GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
export const getAutoSavingAccountTokenConfig = async ({ client, account, token, }) => {
    try {
        const config = (await client.readContract({
            address: GLOBAL_CONSTANTS.AUTO_SAVINGS_ADDRESS,
            abi,
            functionName: 'config',
            args: [account.address, token],
        }));
        return config;
    }
    catch {
        throw new Error(`Failed to get config for token ${token}`);
    }
};
