// @rhinestone/module-sdk/_cjs/module/smart-sessions/policies/universal-action-policy/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUniversalActionPolicy = void 0;
const constants_1 = require("./constants");
const viem_1 = require("viem");
const constants_2 = require("../../../../constants");
const getUniversalActionPolicy = (actionConfig) => {
    if (actionConfig.paramRules.rules.length > constants_1.MAX_RULES) {
        throw new Error(`Max number of rules is ${constants_1.MAX_RULES}`);
    }
    return {
        policy: constants_2.GLOBAL_CONSTANTS.UNIVERSAL_ACTION_POLICY_ADDRESS,
        address: constants_2.GLOBAL_CONSTANTS.UNIVERSAL_ACTION_POLICY_ADDRESS,
        initData: (0, viem_1.encodeAbiParameters)([
            {
                components: [
                    {
                        name: 'valueLimitPerUse',
                        type: 'uint256',
                    },
                    {
                        components: [
                            {
                                name: 'length',
                                type: 'uint256',
                            },
                            {
                                components: [
                                    {
                                        name: 'condition',
                                        type: 'uint8',
                                    },
                                    {
                                        name: 'offset',
                                        type: 'uint64',
                                    },
                                    {
                                        name: 'isLimited',
                                        type: 'bool',
                                    },
                                    {
                                        name: 'ref',
                                        type: 'bytes32',
                                    },
                                    {
                                        components: [
                                            {
                                                name: 'limit',
                                                type: 'uint256',
                                            },
                                            {
                                                name: 'used',
                                                type: 'uint256',
                                            },
                                        ],
                                        name: 'usage',
                                        type: 'tuple',
                                    },
                                ],
                                name: 'rules',
                                type: 'tuple[16]',
                            },
                        ],
                        name: 'paramRules',
                        type: 'tuple',
                    },
                ],
                name: 'ActionConfig',
                type: 'tuple',
            },
        ], [actionConfig]),
    };
};
exports.getUniversalActionPolicy = getUniversalActionPolicy;
