// @rhinestone/module-sdk/_esm/account/kernel/api/installModule.js
import { encodeAbiParameters, encodeFunctionData, encodePacked, parseAbi, zeroAddress, } from 'viem';
import { isModuleInstalled } from './isModuleInstalled.js';
import { accountAbi } from '../constants/abis.js';
import { kernelModuleTypeIds } from '../types.js';
export const installModule = ({ client, account, module, }) => {
    switch (module.type) {
        case 'validator':
        case 'executor':
            return _installModule({ client, account, module, withHook: true });
        case 'hook':
        case 'policy':
        case 'signer':
            return _installModule({ client, account, module, withHook: false });
        case 'fallback':
            return installFallback({ client, account, module });
        default:
            throw new Error(`Unknown module type ${module.type}`);
    }
};
const _installModule = async ({ client, account, module, withHook = false, }) => {
    const executions = [];
    const isInstalled = await isModuleInstalled({ client, account, module });
    if (!isInstalled) {
        const data = encodeFunctionData({
            functionName: 'installModule',
            abi: parseAbi(accountAbi),
            args: [
                BigInt(kernelModuleTypeIds[module.type]),
                module.module,
                withHook
                    ? encodePacked(['address', 'bytes'], [
                        module.hook ?? zeroAddress,
                        encodeAbiParameters([{ type: 'bytes' }, { type: 'bytes' }], [module.initData || '0x', '0x']),
                    ])
                    : module.initData || '0x',
            ],
        });
        executions.push({
            to: account.address,
            target: account.address,
            value: BigInt(0),
            callData: data,
            data,
        });
    }
    return executions;
};
async function installFallback({ client, account, module, }) {
    if (!module.selector || !module.callType) {
        throw new Error(`Hook, selector and callType are required for module type ${module.type}`);
    }
    const executions = [];
    const isInstalled = await isModuleInstalled({
        client,
        account,
        module,
    });
    if (!isInstalled) {
        const data = encodeFunctionData({
            functionName: 'installModule',
            abi: parseAbi(accountAbi),
            args: [
                BigInt(kernelModuleTypeIds[module.type]),
                module.module,
                encodePacked(['bytes4', 'address', 'bytes'], [
                    module.selector,
                    module.hook ?? zeroAddress,
                    encodeAbiParameters([{ type: 'bytes' }, { type: 'bytes' }], [
                        encodePacked(['bytes1', 'bytes'], [module.callType, module.initData || '0x']),
                        '0x',
                    ]),
                ]),
            ],
        });
        executions.push({
            to: account.address,
            target: account.address,
            value: BigInt(0),
            callData: data,
            data,
        });
    }
    return executions;
}
