// @rhinestone/module-sdk/_cjs/module/scheduled-transfers/usage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getToggleScheduledTransferAction = exports.getExecuteScheduledTransferAction = exports.getCreateScheduledTransferAction = exports.getScheduledTransferData = void 0;
const viem_1 = require("viem");
const abi_1 = require("./abi");
const constants_1 = require("../../constants");
const getScheduledTransferData = ({ scheduledTransfer, }) => {
    const amount = BigInt(Number(scheduledTransfer.amount) *
        10 ** (scheduledTransfer.token?.decimals || 18));
    return (0, viem_1.encodeAbiParameters)([
        { name: 'recipient', type: 'address' },
        { name: 'token', type: 'address' },
        { name: 'amount', type: 'uint256' },
    ], [
        scheduledTransfer.recipient,
        scheduledTransfer.token
            ? scheduledTransfer.token.token_address
            : viem_1.zeroAddress,
        amount,
    ]);
};
exports.getScheduledTransferData = getScheduledTransferData;
const getCreateScheduledTransferAction = ({ scheduledTransfer, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'addOrder',
        abi: abi_1.abi,
        args: [
            (0, viem_1.encodePacked)(['uint48', 'uint16', 'uint48', 'bytes'], [
                scheduledTransfer.repeatEvery,
                scheduledTransfer.numberOfRepeats,
                scheduledTransfer.startDate,
                (0, exports.getScheduledTransferData)({ scheduledTransfer }),
            ]),
        ],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        target: constants_1.GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
exports.getCreateScheduledTransferAction = getCreateScheduledTransferAction;
const getExecuteScheduledTransferAction = ({ jobId, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'executeOrder',
        abi: abi_1.abi,
        args: [BigInt(jobId)],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        target: constants_1.GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
exports.getExecuteScheduledTransferAction = getExecuteScheduledTransferAction;
const getToggleScheduledTransferAction = ({ jobId, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'toggleOrder',
        abi: abi_1.abi,
        args: [BigInt(jobId)],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        target: constants_1.GLOBAL_CONSTANTS.SCHEDULED_TRANSFERS_EXECUTOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
exports.getToggleScheduledTransferAction = getToggleScheduledTransferAction;
