// @rhinestone/module-sdk/_cjs/module/multi-factor-validator/installation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMultiFactorValidator = void 0;
const viem_1 = require("viem");
const constants_1 = require("../../constants");
const getMultiFactorValidator = ({ threshold, validators, hook, }) => {
    return {
        address: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
        module: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
        initData: (0, viem_1.encodePacked)(['uint8', 'bytes'], [
            threshold,
            (0, viem_1.encodeAbiParameters)([
                {
                    components: [
                        {
                            internalType: 'bytes32',
                            name: 'packedValidatorAndId',
                            type: 'bytes32',
                        },
                        { internalType: 'bytes', name: 'data', type: 'bytes' },
                    ],
                    name: 'validators',
                    type: 'tuple[]',
                },
            ], [validators]),
        ]),
        deInitData: '0x',
        additionalContext: '0x',
        type: 'validator',
        hook,
    };
};
exports.getMultiFactorValidator = getMultiFactorValidator;
