// @rhinestone/module-sdk/_cjs/module/multi-factor-validator/usage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMFAValidatorMockSignature = exports.isMFASubValidator = exports.getRemoveMFAValidatorAction = exports.getSetMFAValidatorAction = exports.getSetMFAThresholdAction = void 0;
const viem_1 = require("viem");
const abi_1 = require("./abi");
const constants_1 = require("../../constants");
const getSetMFAThresholdAction = ({ threshold, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'setThreshold',
        abi: abi_1.abi,
        args: [threshold],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
        target: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
exports.getSetMFAThresholdAction = getSetMFAThresholdAction;
const getSetMFAValidatorAction = ({ validatorAddress, validatorId, newValidatorData, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'setValidator',
        abi: abi_1.abi,
        args: [validatorAddress, validatorId, newValidatorData],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
        target: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
exports.getSetMFAValidatorAction = getSetMFAValidatorAction;
const getRemoveMFAValidatorAction = ({ validatorAddress, validatorId, }) => {
    const data = (0, viem_1.encodeFunctionData)({
        functionName: 'removeValidator',
        abi: abi_1.abi,
        args: [validatorAddress, validatorId],
    });
    return {
        to: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
        target: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
        value: BigInt(0),
        callData: data,
        data,
    };
};
exports.getRemoveMFAValidatorAction = getRemoveMFAValidatorAction;
const isMFASubValidator = async ({ account, client, subValidator, validatorId, }) => {
    try {
        return (await client.readContract({
            address: constants_1.GLOBAL_CONSTANTS.MULTI_FACTOR_VALIDATOR_ADDRESS,
            abi: abi_1.abi,
            functionName: 'isSubValidator',
            args: [account.address, subValidator, validatorId],
        }));
    }
    catch (err) {
        throw new Error(`Failed to check if ${subValidator} is a sub-validator`);
    }
};
exports.isMFASubValidator = isMFASubValidator;
const getMFAValidatorMockSignature = () => {
    const mockValidators = [
        {
            packedValidatorAndId: (0, viem_1.encodePacked)(['bytes12', 'address'], [
                '0x000000000000000000000000',
                '0xf83d07238a7c8814a48535035602123ad6dbfa63',
            ]),
            data: '0xe8b94748580ca0b4993c9a1b86b5be851bfc076ff5ce3a1ff65bf16392acfcb800f9b4f1aef1555c7fce5599fffb17e7c635502154a0333ba21f3ae491839af51c',
        },
    ];
    return (0, viem_1.encodeAbiParameters)([
        {
            components: [
                {
                    internalType: 'bytes32',
                    name: 'packedValidatorAndId',
                    type: 'bytes32',
                },
                { internalType: 'bytes', name: 'data', type: 'bytes' },
            ],
            name: 'validators',
            type: 'tuple[]',
        },
    ], [mockValidators]);
};
exports.getMFAValidatorMockSignature = getMFAValidatorMockSignature;
