// @rhinestone/module-sdk/_cjs/account/nexus/api/uninstallModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uninstallModule = void 0;
const viem_1 = require("viem");
const types_1 = require("../../../module/types");
const isModuleInstalled_1 = require("./isModuleInstalled");
const abis_1 = require("../constants/abis");
const common_1 = require("../../../common");
const uninstallModule = ({ client, account, module, }) => {
    switch (module.type) {
        case 'validator':
        case 'executor':
        case 'hook':
            return _uninstallModule({ client, account, module });
        case 'fallback':
            if (!module.selector) {
                throw new Error(`Selector param is required for module type ${module.type}`);
            }
            return _uninstallFallback({ client, account, module });
        default:
            throw new Error(`Unknown module type ${module.type}`);
    }
};
exports.uninstallModule = uninstallModule;
const _uninstallModule = async ({ client, account, module, }) => {
    const executions = [];
    const isInstalled = await (0, isModuleInstalled_1.isModuleInstalled)({ client, account, module });
    if (isInstalled) {
        let moduleData = module.deInitData;
        if (module.type === 'validator' || module.type === 'executor') {
            const prev = await (0, common_1.getPreviousModule)({ client, account, module });
            moduleData = (0, viem_1.encodeAbiParameters)([
                { name: 'prev', type: 'address' },
                { name: 'disableModuleData', type: 'bytes' },
            ], [prev, moduleData]);
        }
        const data = (0, viem_1.encodeFunctionData)({
            functionName: 'uninstallModule',
            abi: (0, viem_1.parseAbi)(abis_1.accountAbi),
            args: [BigInt(types_1.moduleTypeIds[module.type]), module.module, moduleData],
        });
        executions.push({
            to: account.address,
            target: account.address,
            value: BigInt(0),
            callData: data,
            data,
        });
    }
    return executions;
};
const _uninstallFallback = async ({ client, account, module, }) => {
    const executions = [];
    const isInstalled = await (0, isModuleInstalled_1.isModuleInstalled)({
        client,
        account,
        module: {
            ...module,
            additionalContext: (0, viem_1.encodeAbiParameters)([{ name: 'functionSignature', type: 'bytes4' }], [module.selector]),
        },
    });
    if (isInstalled) {
        const data = (0, viem_1.encodeFunctionData)({
            functionName: 'uninstallModule',
            abi: (0, viem_1.parseAbi)(abis_1.accountAbi),
            args: [
                BigInt(types_1.moduleTypeIds[module.type]),
                module.module,
                (0, viem_1.encodePacked)(['bytes4', 'bytes'], [module.selector, module.deInitData]),
            ],
        });
        executions.push({
            to: account.address,
            target: account.address,
            value: BigInt(0),
            callData: data,
            data,
        });
    }
    return executions;
};
